--- openbsd_usb.c.orig	Wed Jun  3 15:22:32 2015
+++ openbsd_usb.c	Sat Jun 20 00:11:39 2015
@@ -628,6 +628,7 @@
 	struct libusb_control_setup *setup;
 	struct device_priv *dpriv;
 	struct usb_ctl_request req;
+	struct ctl_urb *urb;
 
 	transfer = USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);
 	dpriv = (struct device_priv *)transfer->dev_handle->dev->os_priv;
@@ -669,11 +670,25 @@
 		}
 		close(fd);
 	} else {
+		urb = malloc(sizeof(*urb));
+		urb->req = req;
+		urb->user_context = itransfer;
+
 		if ((ioctl(dpriv->fd, USB_SET_TIMEOUT, &transfer->timeout)) < 0)
 			return _errno_to_libusb(errno);
 
-		if ((ioctl(dpriv->fd, USB_DO_REQUEST, &req)) < 0)
+		if ((ioctl(dpriv->fd, USB_DO_REQUEST, &urb)) < 0)
 			return _errno_to_libusb(errno);
+
+		while ((ioctl(dpriv->fd, USB_GET_COMPLETED, &urb)) < 0) {}
+
+		itransfer->transferred = urb->req.ucr_actlen;
+
+		usbi_dbg("transferred %d", itransfer->transferred);
+
+		free(urb);
+
+		return (0);
 	}
 
 	itransfer->transferred = req.ucr_actlen;
