--- ugen.c.orig	Wed Jun  3 14:35:32 2015
+++ ugen.c	Mon Jun 15 14:50:01 2015
@@ -104,6 +104,8 @@
 	u_char sc_secondary;
 };
 
+void ugen_request_async_callback(struct usbd_xfer *, void *, usbd_status);
+
 void ugenintr(struct usbd_xfer *xfer, void *addr, usbd_status status);
 void ugen_isoc_rintr(struct usbd_xfer *xfer, void *addr, usbd_status status);
 int ugen_do_read(struct ugen_softc *, int, struct uio *, int);
@@ -130,6 +132,39 @@
 	sizeof(struct ugen_softc), ugen_match, ugen_attach, ugen_detach
 };
 
+struct ugen_info {
+	struct uio	uio;
+	void	       *ptr;
+	int		length;
+};
+
+struct xfer_record {
+        struct usbd_xfer	       *xfer;
+        TAILQ_ENTRY(xfer_record)	next;
+};
+
+TAILQ_HEAD(, xfer_record) xfer_record_head = TAILQ_HEAD_INITIALIZER(xfer_record_head);
+
+void ugen_request_async_callback(struct usbd_xfer *xfer, void *priv, usbd_status s) {
+        struct ugen_info *info = priv;
+	struct xfer_record *xfer_r;
+        int error = 0;
+
+        if (s == USBD_NORMAL_COMPLETION) {
+                if (info->uio.uio_rw == UIO_READ) {
+                        error = uiomovei(info->ptr, info->length, &info->uio);
+                        if (error)
+                                goto ret;
+                }
+
+		xfer_r = malloc(sizeof(struct xfer_record), M_TEMP, M_WAITOK);
+		TAILQ_INSERT_TAIL(&xfer_record_head, xfer_r, next);
+        }
+ret:
+	free(info, M_TEMP, sizeof(*info));
+	usbd_free_xfer(xfer);
+}
+
 int
 ugen_match(struct device *parent, void *match, void *aux)
 {
@@ -393,6 +428,7 @@
 		}
 	}
 	sc->sc_is_open[endpt] = 1;
+	// xfer_record_head = TAILQ_HEAD_INITIALIZER(xfer_record_head);
 	return (0);
 }
 
@@ -455,7 +491,10 @@
 		}
 	}
 	sc->sc_is_open[endpt] = 0;
-
+	// while ((np = TAILQ_FIRST(&xfer_r_head))) {
+	//	TAILQ_REMOVE(&xfer_r_head, np, next);
+	//	free(np);
+	// }
 	return (0);
 }
 
@@ -1144,6 +1183,8 @@
 	case USB_DO_REQUEST:
 	{
 		struct usb_ctl_request *ur = (void *)addr;
+		struct usbd_xfer *xfer;
+		struct ugen_info *info;
 		int len = UGETW(ur->ucr_request.wLength);
 		struct iovec iov;
 		struct uio uio;
@@ -1163,6 +1204,12 @@
 
 		if (len < 0 || len > 32767)
 			return (EINVAL);
+
+		xfer = usbd_alloc_xfer(sc->sc_udev);
+
+		if (xfer == NULL)
+			return (USBD_NOMEM);
+
 		if (len != 0) {
 			iov.iov_base = (caddr_t)ur->ucr_data;
 			iov.iov_len = len;
@@ -1175,34 +1222,60 @@
 				ur->ucr_request.bmRequestType & UT_READ ?
 				UIO_READ : UIO_WRITE;
 			uio.uio_procp = p;
-			ptr = malloc(len, M_TEMP, M_WAITOK);
+			ptr = usbd_alloc_buffer(xfer, len);
+			if (ptr == 0) {
+				usbd_free_xfer(xfer);
+				return (0);
+			}
 			if (uio.uio_rw == UIO_WRITE) {
 				error = uiomovei(ptr, len, &uio);
-				if (error)
-					goto ret;
+				if (error) {
+					usbd_free_xfer(xfer);
+					return (error);
+				}
 			}
 		}
+
 		sce = &sc->sc_endpoints[endpt][IN];
-		err = usbd_do_request_flags(sc->sc_udev, &ur->ucr_request,
-			  ptr, ur->ucr_flags, &ur->ucr_actlen, sce->timeout);
+
+		/* but usbd_request_async calls
+		 * usbd_setup_default_xfer, and that call sets buffer to
+		 * null
+		 */
+		usbd_setup_xfer(xfer, sce->pipeh, NULL, ptr,
+		    len, ur->ucr_flags, sce->timeout, (usbd_callback) ugen_request_async_callback);
+
+		info = malloc(sizeof(*info), M_TEMP, M_NOWAIT);
+		if (info == NULL) {
+			usbd_free_xfer(xfer);
+	                return (-1);
+		}
+
+		info->uio = uio;
+		info->ptr = ptr;
+		info->length = len;
+
+		err = usbd_request_async(xfer, &ur->ucr_request, &info, (usbd_callback) ugen_request_async_callback);
+
 		if (err) {
 			error = EIO;
-			goto ret;
+			free(info, M_TEMP, sizeof(*info));
+			usbd_free_xfer(xfer);
+			return (error);
 		}
-		/* Only if USBD_SHORT_XFER_OK is set. */
-		if (len > ur->ucr_actlen)
-			len = ur->ucr_actlen;
-		if (len != 0) {
-			if (uio.uio_rw == UIO_READ) {
-				error = uiomovei(ptr, len, &uio);
-				if (error)
-					goto ret;
-			}
-		}
-	ret:
-		if (ptr)
-			free(ptr, M_TEMP, 0);
-		return (error);
+		return (0);
+	}
+	case USB_GET_COMPLETED:
+	{
+		/* transfers were allocated in the driver.
+		 * If we were like linux, we'd send back the urb that was
+		 * allocated in user space which now has its buffer filled
+		 * in
+		 */
+		// xfer_record
+		// return TAILQ_FIRST(xfer_record_head);
+		// TAILQ_REMOVE_HEAD(&xfer_record_head, next);
+		return (0);
 	}
 	case USB_GET_DEVICEINFO:
 		usbd_fill_deviceinfo(sc->sc_udev,
