--- ugen.c.orig	Wed Jun  3 14:35:32 2015
+++ ugen.c	Fri Jun 12 17:37:31 2015
@@ -50,6 +50,7 @@
 #include <dev/usb/usbdi.h>
 #include <dev/usb/usbdi_util.h>
 #include <dev/usb/usbdevs.h>
+#include <dev/usb/usbdivar.h>
 
 #ifdef UGEN_DEBUG
 #define DPRINTF(x)	do { if (ugendebug) printf x; } while (0)
@@ -104,6 +105,8 @@
 	u_char sc_secondary;
 };
 
+void ugen_request_async_callback(struct usbd_xfer *, void *, usbd_status);
+
 void ugenintr(struct usbd_xfer *xfer, void *addr, usbd_status status);
 void ugen_isoc_rintr(struct usbd_xfer *xfer, void *addr, usbd_status status);
 int ugen_do_read(struct ugen_softc *, int, struct uio *, int);
@@ -130,6 +133,12 @@
 	sizeof(struct ugen_softc), ugen_match, ugen_attach, ugen_detach
 };
 
+void ugen_request_async_callback(struct usbd_xfer *xfer, void *priv, usbd_status s) {
+	if (xfer->buffer)
+		free(xfer->buffer, M_TEMP, 0);
+	usbd_free_xfer(xfer);
+}
+
 int
 ugen_match(struct device *parent, void *match, void *aux)
 {
@@ -1144,6 +1153,7 @@
 	case USB_DO_REQUEST:
 	{
 		struct usb_ctl_request *ur = (void *)addr;
+		struct usbd_xfer *xfer;
 		int len = UGETW(ur->ucr_request.wLength);
 		struct iovec iov;
 		struct uio uio;
@@ -1182,22 +1192,28 @@
 					goto ret;
 			}
 		}
+
+		usbd_alloc_buffer(struct usbd_xfer *xfer, u_int32_t size)
+
+		xfer = usbd_alloc_xfer(sc->sc_udev);
+
+		if (xfer == NULL)
+			return (USBD_NOMEM);
+
 		sce = &sc->sc_endpoints[endpt][IN];
-		err = usbd_do_request_flags(sc->sc_udev, &ur->ucr_request,
-			  ptr, ur->ucr_flags, &ur->ucr_actlen, sce->timeout);
+
+		/* but usbd_request_async calls
+		 * usbd_setup_default_xfer, and that call sets buffer to
+		 * null
+		 */
+		usbd_setup_xfer(xfer, sce->pipeh, NULL, ptr,
+		    len, ur->ucr_flags, sce->timeout, (usbd_callback) ugen_request_async_callback);
+
+		err = usbd_request_async(xfer, &ur->ucr_request, &uio, (usbd_callback) ugen_request_async_callback);
+
 		if (err) {
 			error = EIO;
 			goto ret;
-		}
-		/* Only if USBD_SHORT_XFER_OK is set. */
-		if (len > ur->ucr_actlen)
-			len = ur->ucr_actlen;
-		if (len != 0) {
-			if (uio.uio_rw == UIO_READ) {
-				error = uiomovei(ptr, len, &uio);
-				if (error)
-					goto ret;
-			}
 		}
 	ret:
 		if (ptr)
