--- ugen.c.orig	Wed Jun  3 14:35:32 2015
+++ ugen.c	Tue Jun 30 16:28:16 2015
@@ -46,9 +46,13 @@
 #include <sys/vnode.h>
 #include <sys/poll.h>
 
+#include <machine/bus.h>
+
 #include <dev/usb/usb.h>
 #include <dev/usb/usbdi.h>
 #include <dev/usb/usbdi_util.h>
+#include <dev/usb/usbdivar.h>
+#include <dev/usb/usb_mem.h>
 #include <dev/usb/usbdevs.h>
 
 #ifdef UGEN_DEBUG
@@ -104,6 +108,8 @@
 	u_char sc_secondary;
 };
 
+void ugen_request_async_callback(struct usbd_xfer *, void *, usbd_status);
+
 void ugenintr(struct usbd_xfer *xfer, void *addr, usbd_status status);
 void ugen_isoc_rintr(struct usbd_xfer *xfer, void *addr, usbd_status status);
 int ugen_do_read(struct ugen_softc *, int, struct uio *, int);
@@ -130,6 +136,37 @@
 	sizeof(struct ugen_softc), ugen_match, ugen_attach, ugen_detach
 };
 
+TAILQ_HEAD(, ctl_urb) urb_entry_head;
+
+void ugen_request_async_callback(struct usbd_xfer *xfer, void *priv, usbd_status s) {
+	struct ctl_urb *urb = priv;
+
+        if (s == USBD_NORMAL_COMPLETION || s == USBD_SHORT_XFER) {
+		urb->actlen = xfer->actlen;
+		urb->dmabuf = KERNADDR(&xfer->dmabuf, 0);
+        }
+	urb->status = (int)s;
+	urb->xfer = xfer;
+
+	TAILQ_INSERT_TAIL(&urb_entry_head, urb, entries);
+}
+
+void ugen_read_async_callback(struct usbd_xfer *xfer, void *priv, usbd_status s) {
+	struct ctl_urb *urb = priv;
+	/* more transfers left? otherwise, insert kurb onto
+	 * completeded queue */
+	printf("read callback called\n");
+        if (s == USBD_NORMAL_COMPLETION || s == USBD_SHORT_XFER) {
+		printf("normal completion\n");
+		urb->actlen = xfer->actlen;
+		urb->dmabuf = KERNADDR(&xfer->dmabuf, 0);
+        }
+	urb->status = (int)s;
+	urb->xfer = xfer;
+
+	TAILQ_INSERT_TAIL(&urb_entry_head, urb, entries);
+}
+
 int
 ugen_match(struct device *parent, void *match, void *aux)
 {
@@ -150,6 +187,8 @@
 	usbd_status err;
 	int conf;
 
+	TAILQ_INIT(&urb_entry_head);
+
 	sc->sc_udev = udev = uaa->device;
 
 	if (usbd_get_devcnt(udev) > 0)
@@ -986,6 +1025,102 @@
 			return (EINVAL);
 		sce->timeout = *(int *)addr;
 		return (0);
+	case USB_ASYNC_SUBMIT:
+		struct ctl_urb *urb = (void *)addr;
+		int flags = 0;
+		int left = 0;
+
+		if (urb->length > UGEN_BBSIZE)
+			return EIO;
+		printf("length ok\n");
+
+		xfer = usbd_alloc_xfer(sc->sc_udev);
+		if (xfer == 0)
+			return (ENOMEM);
+		kurb = malloc(sizeof(*kurb), M_TEMP, M_WAITOK);
+		if (kurb == NULL) {
+			usbd_free_xfer(xfer);
+			return (ENOMEM);
+		}
+		*kurb = *urb;
+		buf = usbd_alloc_buffer(xfer, kurb->length);
+		if (buf == NULL) {
+			usbd_free_xfer(xfer);
+			free(kurb, M_TEMP, sizeof(*kurb));
+			return (ENOMEM);
+		}
+
+		sce = &sc->sc_endpoints[endpt][IN];
+		if (sce->state & UGEN_SHORT_OK)
+			flags = USBD_SHORT_XFER_OK;
+		usbd_setup_xfer(xfer, sce->pipeh, kurb, buf, urb->length,
+		    flags, sce->timeout, (usbd_callback) ugen_read_async_callback);
+		err = usbd_transfer(xfer);
+		printf("transfer submitted\n");
+		if (err) {
+			printf("transfer error\n");
+			usbd_clear_endpoint_stall(sce->pipeh);
+			if (err == USBD_INTERRUPTED)
+				error = EINTR;
+			else if (err == USBD_TIMEOUT)
+				error = ETIMEDOUT;
+			else
+				error = EIO;
+			usbd_free_xfer(xfer);
+			free(kurb, M_TEMP, sizeof(*kurb));
+			return (error);
+		}
+		printf("submit success\n");
+		return (0);
+	case USB_ASYNC_COMPLETE:
+		struct ctl_urb *urb = (void *)addr;
+		struct ctl_urb *kurb;
+		int flags = 0;
+		struct uio uio;
+		struct iovec iov;
+		int s;
+		int error = 0;
+
+		printf("GETTING COMPLETED\n");
+		s = splusb();
+		kurb = TAILQ_FIRST(&urb_entry_head);
+		if (kurb == NULL) {
+			printf("nothing\n");
+			splx(s);
+			return (EIO);
+		}
+		TAILQ_REMOVE(&urb_entry_head, kurb, entries);
+		splx(s);
+
+		printf("got one\n");
+		if (kurb->status == USBD_NORMAL_COMPLETION ||
+		    kurb->status == USBD_SHORT_XFER) {
+			printf("it was successful\n");
+			/* construct uio from ctl_urb */
+			if (kurb->actlen != 0) {
+				iov.iov_base = (caddr_t)kurb->buffer;
+				iov.iov_len = kurb->actlen;
+				uio.uio_iov = &iov;
+				uio.uio_iovcnt = 1;
+				uio.uio_resid = kurb->actlen;
+				uio.uio_offset = 0;
+				uio.uio_segflg = UIO_USERSPACE;
+				uio.uio_rw = UIO_READ;
+				uio.uio_procp = p;
+				printf("do move\n");
+				error = uiomovei(kurb->dmabuf, kurb->actlen, &uio);
+				if (error) {
+					printf("move error\n");
+					usbd_free_xfer(kurb->xfer);
+					free(kurb, M_TEMP, sizeof(*kurb));
+					return (error);
+				}
+			}
+		}
+		*urb = *kurb;
+		usbd_free_xfer(kurb->xfer);
+		free(kurb, M_TEMP, sizeof(*kurb));
+		return (0);
 	default:
 		break;
 	}
@@ -1143,11 +1278,14 @@
 	}
 	case USB_DO_REQUEST:
 	{
-		struct usb_ctl_request *ur = (void *)addr;
+		struct ctl_urb *urb = (void *)addr;
+		struct usb_ctl_request *ur = &urb->req;
 		int len = UGETW(ur->ucr_request.wLength);
+		struct ctl_urb *kurb;
+		struct usbd_xfer *xfer;
+		void *buf;
 		struct iovec iov;
 		struct uio uio;
-		void *ptr = 0;
 		int error = 0;
 
 		if (!(flag & FWRITE))
@@ -1160,9 +1298,18 @@
 		    (ur->ucr_request.bmRequestType == UT_WRITE_INTERFACE &&
 		     ur->ucr_request.bRequest == UR_SET_INTERFACE))
 			return (EINVAL);
-
 		if (len < 0 || len > 32767)
 			return (EINVAL);
+
+		xfer = usbd_alloc_xfer(sc->sc_udev);
+		if (xfer == NULL)
+			return (ENOMEM);
+		kurb = malloc(sizeof(*kurb), M_TEMP, M_WAITOK);
+		if (kurb == NULL) {
+			usbd_free_xfer(xfer);
+			return (ENOMEM);
+		}
+		*kurb = *urb;
 		if (len != 0) {
 			iov.iov_base = (caddr_t)ur->ucr_data;
 			iov.iov_len = len;
@@ -1175,34 +1322,84 @@
 				ur->ucr_request.bmRequestType & UT_READ ?
 				UIO_READ : UIO_WRITE;
 			uio.uio_procp = p;
-			ptr = malloc(len, M_TEMP, M_WAITOK);
+			buf = usbd_alloc_buffer(xfer, len);
+			if (buf == NULL) {
+				usbd_free_xfer(xfer);
+				free(kurb, M_TEMP, sizeof(*kurb));
+				return (ENOMEM);
+			}
 			if (uio.uio_rw == UIO_WRITE) {
-				error = uiomovei(ptr, len, &uio);
-				if (error)
-					goto ret;
+				error = uiomovei(buf, len, &uio);
+				if (error) {
+					usbd_free_xfer(xfer);
+					free(kurb, M_TEMP, sizeof(*kurb));
+					return (error);
+				}
 			}
 		}
-		sce = &sc->sc_endpoints[endpt][IN];
-		err = usbd_do_request_flags(sc->sc_udev, &ur->ucr_request,
-			  ptr, ur->ucr_flags, &ur->ucr_actlen, sce->timeout);
-		if (err) {
-			error = EIO;
-			goto ret;
+
+		error = usbd_request_async(xfer,
+		    &ur->ucr_request, kurb, (usbd_callback)
+		    ugen_request_async_callback);
+
+		if (error) {
+			free(kurb, M_TEMP, sizeof(*kurb));
+			usbd_free_xfer(xfer);
+			return (error);
 		}
-		/* Only if USBD_SHORT_XFER_OK is set. */
-		if (len > ur->ucr_actlen)
-			len = ur->ucr_actlen;
-		if (len != 0) {
-			if (uio.uio_rw == UIO_READ) {
-				error = uiomovei(ptr, len, &uio);
-				if (error)
-					goto ret;
+		break;
+	}
+	case USB_GET_COMPLETED:
+	{
+		struct ctl_urb *urb = (struct ctl_urb *)addr;
+		struct ctl_urb *kurb;
+		struct usb_ctl_request *ur;
+		int len;
+		struct iovec iov;
+		struct uio uio;
+		int s;
+		int error = 0;
+
+		s = splusb();
+		kurb = TAILQ_FIRST(&urb_entry_head);
+		if (kurb == NULL) {
+			splx(s);
+			return (EIO);
+		}
+		TAILQ_REMOVE(&urb_entry_head, kurb, entries);
+		splx(s);
+		if (kurb->status == USBD_NORMAL_COMPLETION ||
+		    kurb->status == USBD_SHORT_XFER) {
+			ur = &kurb->req;
+			len = UGETW(ur->ucr_request.wLength);
+			if (len > kurb->actlen)
+				len = kurb->actlen;
+			if (len != 0) {
+				iov.iov_base = (caddr_t)ur->ucr_data;
+				iov.iov_len = len;
+				uio.uio_iov = &iov;
+				uio.uio_iovcnt = 1;
+				uio.uio_resid = len;
+				uio.uio_offset = 0;
+				uio.uio_segflg = UIO_USERSPACE;
+				uio.uio_rw =
+					ur->ucr_request.bmRequestType & UT_READ ?
+					UIO_READ : UIO_WRITE;
+				uio.uio_procp = p;
+				if (uio.uio_rw == UIO_READ) {
+					error = uiomove(kurb->dmabuf, len, &uio);
+					if (error) {
+						usbd_free_xfer(kurb->xfer);
+						free(kurb, M_TEMP, sizeof(*kurb));
+						return (error);
+					}
+				}
 			}
 		}
-	ret:
-		if (ptr)
-			free(ptr, M_TEMP, 0);
-		return (error);
+		*urb = *kurb;
+		usbd_free_xfer(kurb->xfer);
+		free(kurb, M_TEMP, sizeof(*kurb));
+		break;
 	}
 	case USB_GET_DEVICEINFO:
 		usbd_fill_deviceinfo(sc->sc_udev,
@@ -1248,44 +1445,55 @@
 	if (sce == NULL)
 		return (POLLERR);
 #ifdef DIAGNOSTIC
-	if (!sce->edesc) {
-		printf("ugenpoll: no edesc\n");
-		return (POLLERR);
+	if (UGENENDPOINT(dev) != USB_CONTROL_ENDPOINT) {
+		if (!sce->edesc) {
+			printf("ugenpoll: no edesc\n");
+			return (POLLERR);
+		}
+		if (!sce->pipeh) {
+			printf("ugenpoll: no pipe\n");
+			return (POLLERR);
+		}
 	}
-	if (!sce->pipeh) {
-		printf("ugenpoll: no pipe\n");
-		return (POLLERR);
-	}
 #endif
 	s = splusb();
-	switch (sce->edesc->bmAttributes & UE_XFERTYPE) {
-	case UE_INTERRUPT:
+	if (UGENENDPOINT(dev) == USB_CONTROL_ENDPOINT) {
 		if (events & (POLLIN | POLLRDNORM)) {
-			if (sce->q.c_cc > 0)
+			if (!TAILQ_EMPTY(&urb_entry_head))
 				revents |= events & (POLLIN | POLLRDNORM);
 			else
 				selrecord(p, &sce->rsel);
 		}
-		break;
-	case UE_ISOCHRONOUS:
-		if (events & (POLLIN | POLLRDNORM)) {
-			if (sce->cur != sce->fill)
-				revents |= events & (POLLIN | POLLRDNORM);
-			else
-				selrecord(p, &sce->rsel);
+	} else {
+		switch (sce->edesc->bmAttributes & UE_XFERTYPE) {
+		case UE_INTERRUPT:
+			if (events & (POLLIN | POLLRDNORM)) {
+				if (sce->q.c_cc > 0)
+					revents |= events & (POLLIN | POLLRDNORM);
+				else
+					selrecord(p, &sce->rsel);
+			}
+			break;
+		case UE_ISOCHRONOUS:
+			if (events & (POLLIN | POLLRDNORM)) {
+				if (sce->cur != sce->fill)
+					revents |= events & (POLLIN | POLLRDNORM);
+				else
+					selrecord(p, &sce->rsel);
+			}
+			break;
+		case UE_BULK:
+			/*
+			 * We have no easy way of determining if a read will
+			 * yield any data or a write will happen.
+			 * Pretend they will.
+			 */
+			revents |= events &
+				   (POLLIN | POLLRDNORM | POLLOUT | POLLWRNORM);
+			break;
+		default:
+			break;
 		}
-		break;
-	case UE_BULK:
-		/*
-		 * We have no easy way of determining if a read will
-		 * yield any data or a write will happen.
-		 * Pretend they will.
-		 */
-		revents |= events &
-			   (POLLIN | POLLRDNORM | POLLOUT | POLLWRNORM);
-		break;
-	default:
-		break;
 	}
 	splx(s);
 	return (revents);
