--- ugen.c.orig	Wed Jun  3 14:35:32 2015
+++ ugen.c	Tue Jul  7 04:06:02 2015
@@ -1,4 +1,4 @@
-/*	$OpenBSD: ugen.c,v 1.81 2015/02/10 21:56:09 miod Exp $ */
+/*	$OpenBSD: ugen.c,v 1.84 2015/06/15 15:45:28 mpi Exp $ */
 /*	$NetBSD: ugen.c,v 1.63 2002/11/26 18:49:48 christos Exp $	*/
 /*	$FreeBSD: src/sys/dev/usb/ugen.c,v 1.26 1999/11/17 22:33:41 n_hibma Exp $	*/
 
@@ -46,9 +46,13 @@
 #include <sys/vnode.h>
 #include <sys/poll.h>
 
+#include <machine/bus.h>
+
 #include <dev/usb/usb.h>
 #include <dev/usb/usbdi.h>
 #include <dev/usb/usbdi_util.h>
+#include <dev/usb/usbdivar.h>
+#include <dev/usb/usb_mem.h>
 #include <dev/usb/usbdevs.h>
 
 #ifdef UGEN_DEBUG
@@ -89,6 +93,7 @@
 		void *dmabuf;
 		u_int16_t sizes[UGEN_NISORFRMS];
 	} isoreqs[UGEN_NISOREQS];
+	TAILQ_HEAD(, ctl_urb) queue;
 };
 
 struct ugen_softc {
@@ -104,12 +109,15 @@
 	u_char sc_secondary;
 };
 
+void ugen_request_async_callback(struct usbd_xfer *, void *, usbd_status);
+
 void ugenintr(struct usbd_xfer *xfer, void *addr, usbd_status status);
 void ugen_isoc_rintr(struct usbd_xfer *xfer, void *addr, usbd_status status);
 int ugen_do_read(struct ugen_softc *, int, struct uio *, int);
 int ugen_do_write(struct ugen_softc *, int, struct uio *, int);
-int ugen_do_ioctl(struct ugen_softc *, int, u_long,
-			 caddr_t, int, struct proc *);
+int ugen_do_ioctl(struct ugen_softc *, int, u_long, caddr_t, int,
+	struct proc *);
+int ugen_do_close(struct ugen_softc *, int, int);
 int ugen_set_config(struct ugen_softc *sc, int configno);
 int ugen_set_interface(struct ugen_softc *, int, int);
 int ugen_get_alt_index(struct ugen_softc *sc, int ifaceidx);
@@ -130,6 +138,32 @@
 	sizeof(struct ugen_softc), ugen_match, ugen_attach, ugen_detach
 };
 
+void ugen_request_async_callback(struct usbd_xfer *xfer, void *priv, usbd_status s) {
+	struct ctl_urb *urb = priv;
+	struct ugen_endpoint *sce = (struct ugen_endpoint *)urb->sce;
+
+	urb->status = (int)s;
+	urb->xfer = xfer;
+
+	TAILQ_INSERT_TAIL(&sce->queue, urb, entries);
+	selwakeup(&sce->rsel);
+}
+
+void ugen_read_async_callback(struct usbd_xfer *xfer, void *priv, usbd_status s) {
+	struct xfer_w *wrap = priv;
+	struct ctl_urb *urb = wrap->parent;
+	struct ugen_endpoint *sce = (struct ugen_endpoint *)urb->sce;
+
+	wrap->xfer = xfer;
+	urb->count--;
+	TAILQ_INSERT_TAIL(&urb->xfers_head, wrap, entries);
+
+	if (urb->count == 0) {
+		TAILQ_INSERT_TAIL(&sce->queue, urb, entries);
+		selwakeup(&sce->rsel);
+	}
+}
+
 int
 ugen_match(struct device *parent, void *match, void *aux)
 {
@@ -282,6 +316,9 @@
 	if (sc->sc_is_open[endpt])
 		return (EBUSY);
 
+	sce = &sc->sc_endpoints[endpt][IN];
+	TAILQ_INIT(&sce->queue);
+
 	if (endpt == USB_CONTROL_ENDPOINT) {
 		sc->sc_is_open[USB_CONTROL_ENDPOINT] = 1;
 		return (0);
@@ -374,11 +411,10 @@
 				sce->isoreqs[i].dmabuf = buf;
 				for(j = 0; j < UGEN_NISORFRMS; ++j)
 					sce->isoreqs[i].sizes[j] = isize;
-				usbd_setup_isoc_xfer
-					(xfer, sce->pipeh, &sce->isoreqs[i],
-					 sce->isoreqs[i].sizes,
-					 UGEN_NISORFRMS, USBD_NO_COPY,
-					 ugen_isoc_rintr);
+				usbd_setup_isoc_xfer(xfer, sce->pipeh,
+				    &sce->isoreqs[i], sce->isoreqs[i].sizes,
+				    UGEN_NISORFRMS, USBD_NO_COPY |
+				    USBD_SHORT_XFER_OK, ugen_isoc_rintr);
 				(void)usbd_transfer(xfer);
 			}
 			DPRINTFN(5, ("ugenopen: isoc open done\n"));
@@ -399,23 +435,57 @@
 int
 ugenclose(dev_t dev, int flag, int mode, struct proc *p)
 {
+	struct ugen_softc *sc = ugen_cd.cd_devs[UGENUNIT(dev)];
 	int endpt = UGENENDPOINT(dev);
-	struct ugen_softc *sc;
-	struct ugen_endpoint *sce;
-	int dir;
-	int i;
+	int error;
 
-	sc = ugen_cd.cd_devs[UGENUNIT(dev)];
+	if (sc == NULL || usbd_is_dying(sc->sc_udev))
+		return (EIO);
 
 	DPRINTFN(5, ("ugenclose: flag=%d, mode=%d, unit=%d, endpt=%d\n",
 		     flag, mode, UGENUNIT(dev), endpt));
 
+	sc->sc_refcnt++;
+	error = ugen_do_close(sc, endpt, flag);
+	if (--sc->sc_refcnt < 0)
+		usb_detach_wakeup(&sc->sc_dev);
+
+	return (error);
+}
+
+int
+ugen_do_close(struct ugen_softc *sc, int endpt, int flag)
+{
+	struct ugen_endpoint *sce;
+	int dir, i;
+	struct ctl_urb *urb;
+	struct xfer_w *wrap;
+
+
 #ifdef DIAGNOSTIC
 	if (!sc->sc_is_open[endpt]) {
 		printf("ugenclose: not open\n");
 		return (EINVAL);
 	}
 #endif
+	sce = &sc->sc_endpoints[endpt][IN];
+	if (endpt == USB_CONTROL_ENDPOINT) {
+		while ((urb = TAILQ_FIRST(&sce->queue))) {
+			TAILQ_REMOVE(&sce->queue, urb, entries);
+			usbd_free_xfer(urb->xfer);
+			free(urb, M_TEMP, sizeof(*urb));
+		}
+	} else {
+		while ((urb = TAILQ_FIRST(&sce->queue))) {
+			TAILQ_REMOVE(&sce->queue, urb, entries);
+			while ((wrap = TAILQ_FIRST(&urb->xfers_head))) {
+				TAILQ_REMOVE(&urb->xfers_head, wrap, entries);
+				usbd_free_xfer(wrap->xfer);
+				free(wrap, M_TEMP, sizeof(*wrap));
+			}
+			free(urb, M_TEMP, sizeof(*urb));
+		}
+	}
 
 	if (endpt == USB_CONTROL_ENDPOINT) {
 		DPRINTFN(5, ("ugenclose: close control\n"));
@@ -432,7 +502,6 @@
 		DPRINTFN(5, ("ugenclose: endpt=%d dir=%d sce=%p\n",
 			     endpt, dir, sce));
 
-		usbd_abort_pipe(sce->pipeh);
 		usbd_close_pipe(sce->pipeh);
 		sce->pipeh = NULL;
 
@@ -740,9 +809,8 @@
 {
 	struct ugen_softc *sc = (struct ugen_softc *)self;
 	struct ugen_endpoint *sce;
-	int i, dir;
-	int s;
-	int maj, mn;
+	int i, dir, endptno;
+	int s, maj, mn;
 
 	DPRINTF(("ugen_detach: sc=%p flags=%d\n", sc, flags));
 
@@ -774,6 +842,10 @@
 	mn = self->dv_unit * USB_MAX_ENDPOINTS;
 	vdevgone(maj, mn, mn + USB_MAX_ENDPOINTS - 1, VCHR);
 
+	for (endptno = 0; endptno < USB_MAX_ENDPOINTS; endptno++) {
+		if (sc->sc_is_open[endptno])
+			ugen_do_close(sc, endptno, FREAD|FWRITE);
+	}
 	return (0);
 }
 
@@ -859,7 +931,7 @@
 	}
 
 	usbd_setup_isoc_xfer(xfer, sce->pipeh, req, req->sizes, UGEN_NISORFRMS,
-			     USBD_NO_COPY, ugen_isoc_rintr);
+	    USBD_NO_COPY | USBD_SHORT_XFER_OK, ugen_isoc_rintr);
 	(void)usbd_transfer(xfer);
 
 	if (sce->state & UGEN_ASLP) {
@@ -940,8 +1012,8 @@
 }
 
 int
-ugen_do_ioctl(struct ugen_softc *sc, int endpt, u_long cmd,
-	      caddr_t addr, int flag, struct proc *p)
+ugen_do_ioctl(struct ugen_softc *sc, int endpt, u_long cmd, caddr_t addr,
+    int flag, struct proc *p)
 {
 	struct ugen_endpoint *sce;
 	int err;
@@ -986,6 +1058,209 @@
 			return (EINVAL);
 		sce->timeout = *(int *)addr;
 		return (0);
+	case USB_ASYNC_SUBMIT:
+	{
+		struct ctl_urb *urb = (void *)addr;
+		struct usbd_xfer *xfer;
+		struct ctl_urb *kurb;
+		struct xfer_w *wrap;
+		void *buf;
+		int flags = 0;
+		int error = 0;
+		int left = urb->actlen;
+		int len;
+		//struct uio;
+		//struct iovec;
+
+		// specify if read or write
+
+		kurb = malloc(sizeof(*kurb), M_TEMP, M_WAITOK);
+		if (kurb == NULL) {
+			printf("no mem kurb\n");
+			return (ENOMEM);
+		}
+		*kurb = *urb;
+		sce = &sc->sc_endpoints[endpt][IN];
+		if (sce->state & UGEN_SHORT_OK)
+			flags = USBD_SHORT_XFER_OK;
+		flags |= USBD_NO_COPY;
+		kurb->sce = sce;
+
+		kurb->count = left / UGEN_BBSIZE;
+		if (left % UGEN_BBSIZE)
+			kurb->count++;
+
+		TAILQ_INIT(&kurb->xfers_head);
+
+		//if (write) {
+		//	iov.iov_base = (caddr_t)urb->buffer;
+		//	iov.iov_len = urb->actlen;
+		//	uio.uio_iov = &iov;
+		//	uio.uio_iovcnt = 1;
+		//	uio.uio_resid = urb->actlen;
+		//	uio.uio_offset = 0;
+		//	uio.uio_segflg = UIO_USERSPACE;
+		//	uio.uio_rw =
+		//		ur->ucr_request.bmRequestType & UT_READ ?
+		//		UIO_READ : UIO_WRITE;
+		//	uio.uio_procp = p;
+		//}
+
+		while (left != 0) {
+			xfer = usbd_alloc_xfer(sc->sc_udev);
+			if (xfer == 0) {
+				// if no transfers have
+				// been submitted, need to
+				// free kurb here
+
+				// if transfers have been
+				// submitted, should cancel
+				// them here
+				printf("no mem xfer\n");
+				return (ENOMEM);
+			}
+			len = min(UGEN_BBSIZE, left);
+			buf = usbd_alloc_buffer(xfer, len);
+			if (buf == NULL) {
+				// if no transfers have
+				// been submitted, need to
+				// free kurb here
+
+				// if transfers have been
+				// submitted, should cancel
+				// them here
+				printf("no mem buf\n");
+				usbd_free_xfer(xfer);
+				return (ENOMEM);
+			}
+			//if (uio.uio_rw == UIO_WRITE) {
+			//	error = uiomove(buf, len, &uio);
+			//	if (error) {
+			//		usbd_free_xfer(xfer);
+			//		free(kurb, M_TEMP, sizeof(*kurb));
+			//		return (error);
+			//	}
+			//}
+			wrap = malloc(sizeof(*wrap), M_TEMP, M_WAITOK);
+			if (wrap == NULL) {
+				// if no transfers have
+				// been submitted, need to
+				// free kurb here
+
+				// if transfers have been
+				// submitted, should cancel
+				// them here
+				printf("no mem wrap\n");
+				usbd_free_xfer(xfer);
+				return (ENOMEM);
+			}
+			wrap->parent = kurb;
+			left -= len;
+			usbd_setup_xfer(xfer, sce->pipeh, wrap, buf, len,
+			    flags, sce->timeout, (usbd_callback) ugen_read_async_callback);
+			err = usbd_transfer(xfer);
+			if (err != USBD_IN_PROGRESS) {
+				printf("transfer error\n");
+				usbd_clear_endpoint_stall(sce->pipeh);
+				if (err == USBD_INTERRUPTED) {
+					error = EINTR;
+				} else if (err == USBD_TIMEOUT) {
+					error = ETIMEDOUT;
+				} else {
+					error = EIO;
+				}
+				usbd_free_xfer(xfer);
+				free(wrap, M_TEMP, sizeof(*wrap));
+				// if no transfers have
+				// been submitted, need to
+				// free kurb here
+
+				// if transfers have been
+				// submitted, should cancel
+				// them here
+				return (error);
+			}
+		}
+
+		return (0);
+	}
+	case USB_ASYNC_COMPLETE:
+	{
+		struct ctl_urb *urb = (void *)addr;
+		struct ctl_urb *kurb;
+		struct xfer_w *wrap;
+		struct usbd_xfer *xfer;
+		struct uio uio;
+		struct iovec iov;
+		int s;
+		int error = 0;
+
+		sce = &sc->sc_endpoints[endpt][IN];
+
+		s = splusb();
+		kurb = TAILQ_FIRST(&sce->queue);
+		if (kurb == NULL) {
+			splx(s);
+			return (EIO);
+		}
+		TAILQ_REMOVE(&sce->queue, kurb, entries);
+
+		iov.iov_base = (caddr_t)kurb->buffer;
+		iov.iov_len = kurb->actlen;
+		uio.uio_iov = &iov;
+		uio.uio_iovcnt = 1;
+		uio.uio_resid = kurb->actlen;
+		uio.uio_offset = 0;
+		uio.uio_segflg = UIO_USERSPACE;
+		uio.uio_rw = UIO_READ;
+		uio.uio_procp = p;
+
+		while ((wrap = TAILQ_FIRST(&kurb->xfers_head))) {
+			TAILQ_REMOVE(&kurb->xfers_head, wrap, entries);
+			xfer = (struct usbd_xfer *)wrap->xfer;
+			if (xfer->status == USBD_NORMAL_COMPLETION ||
+			    xfer->status == USBD_SHORT_XFER) {
+				error = uiomove(KERNADDR(&xfer->dmabuf, 0), xfer->actlen, &uio);
+				if (error) {
+					printf("move error\n");
+					kurb->status = USBD_IOERROR;
+					*urb = *kurb;
+					usbd_free_xfer(xfer);
+					free(wrap, M_TEMP, sizeof(*wrap));
+					while ((wrap = TAILQ_FIRST(&kurb->xfers_head))) {
+						TAILQ_REMOVE(&kurb->xfers_head, wrap, entries);
+						usbd_free_xfer(wrap->xfer);
+						free(wrap, M_TEMP, sizeof(*wrap));
+					}
+					free(kurb, M_TEMP, sizeof(*kurb));
+					splx(s);
+					return (0);
+				}
+			} else {
+				printf("status not ok\n");
+				printf("status %d\n", (int)xfer->status);
+				kurb->status = USBD_IOERROR;
+				*urb = *kurb;
+				usbd_free_xfer(xfer);
+				free(wrap, M_TEMP, sizeof(*wrap));
+				while ((wrap = TAILQ_FIRST(&kurb->xfers_head))) {
+					TAILQ_REMOVE(&kurb->xfers_head, wrap, entries);
+					usbd_free_xfer(wrap->xfer);
+					free(wrap, M_TEMP, sizeof(*wrap));
+				}
+				free(kurb, M_TEMP, sizeof(*kurb));
+				splx(s);
+				return (0);
+			}
+			usbd_free_xfer(wrap->xfer);
+			free(wrap, M_TEMP, sizeof(*wrap));
+		}
+		splx(s);
+		kurb->status = USBD_NORMAL_COMPLETION;
+		*urb = *kurb;
+		free(kurb, M_TEMP, sizeof(*kurb));
+		return (0);
+	}
 	default:
 		break;
 	}
@@ -1143,11 +1418,14 @@
 	}
 	case USB_DO_REQUEST:
 	{
-		struct usb_ctl_request *ur = (void *)addr;
+		struct ctl_urb *urb = (void *)addr;
+		struct usb_ctl_request *ur = &urb->req;
 		int len = UGETW(ur->ucr_request.wLength);
+		struct ctl_urb *kurb;
+		struct usbd_xfer *xfer;
+		void *buf;
 		struct iovec iov;
 		struct uio uio;
-		void *ptr = 0;
 		int error = 0;
 
 		if (!(flag & FWRITE))
@@ -1160,9 +1438,18 @@
 		    (ur->ucr_request.bmRequestType == UT_WRITE_INTERFACE &&
 		     ur->ucr_request.bRequest == UR_SET_INTERFACE))
 			return (EINVAL);
-
 		if (len < 0 || len > 32767)
 			return (EINVAL);
+
+		xfer = usbd_alloc_xfer(sc->sc_udev);
+		if (xfer == NULL)
+			return (ENOMEM);
+		kurb = malloc(sizeof(*kurb), M_TEMP, M_WAITOK);
+		if (kurb == NULL) {
+			usbd_free_xfer(xfer);
+			return (ENOMEM);
+		}
+		*kurb = *urb;
 		if (len != 0) {
 			iov.iov_base = (caddr_t)ur->ucr_data;
 			iov.iov_len = len;
@@ -1175,34 +1462,92 @@
 				ur->ucr_request.bmRequestType & UT_READ ?
 				UIO_READ : UIO_WRITE;
 			uio.uio_procp = p;
-			ptr = malloc(len, M_TEMP, M_WAITOK);
+			buf = usbd_alloc_buffer(xfer, len);
+			if (buf == NULL) {
+				usbd_free_xfer(xfer);
+				free(kurb, M_TEMP, sizeof(*kurb));
+				return (ENOMEM);
+			}
 			if (uio.uio_rw == UIO_WRITE) {
-				error = uiomovei(ptr, len, &uio);
-				if (error)
-					goto ret;
+				error = uiomove(buf, len, &uio);
+				if (error) {
+					usbd_free_xfer(xfer);
+					free(kurb, M_TEMP, sizeof(*kurb));
+					return (error);
+				}
 			}
 		}
+
 		sce = &sc->sc_endpoints[endpt][IN];
-		err = usbd_do_request_flags(sc->sc_udev, &ur->ucr_request,
-			  ptr, ur->ucr_flags, &ur->ucr_actlen, sce->timeout);
-		if (err) {
-			error = EIO;
-			goto ret;
+		kurb->sce = sce;
+
+		error = usbd_request_async(xfer,
+		    &ur->ucr_request, kurb, (usbd_callback)
+		    ugen_request_async_callback);
+
+		if (error) {
+			free(kurb, M_TEMP, sizeof(*kurb));
+			usbd_free_xfer(xfer);
+			return (error);
 		}
-		/* Only if USBD_SHORT_XFER_OK is set. */
-		if (len > ur->ucr_actlen)
-			len = ur->ucr_actlen;
-		if (len != 0) {
-			if (uio.uio_rw == UIO_READ) {
-				error = uiomovei(ptr, len, &uio);
-				if (error)
-					goto ret;
+		break;
+	}
+	case USB_GET_COMPLETED:
+	{
+		struct ctl_urb *urb = (struct ctl_urb *)addr;
+		struct ctl_urb *kurb;
+		struct usbd_xfer *xfer;
+		struct usb_ctl_request *ur;
+		int len;
+		struct iovec iov;
+		struct uio uio;
+		int s;
+		int error = 0;
+
+		sce = &sc->sc_endpoints[endpt][IN];
+
+		s = splusb();
+		kurb = TAILQ_FIRST(&sce->queue);
+		if (kurb == NULL) {
+			splx(s);
+			return (EIO);
+		}
+		TAILQ_REMOVE(&sce->queue, kurb, entries);
+		splx(s);
+		xfer = (struct usbd_xfer *)kurb->xfer;
+		if (kurb->status == USBD_NORMAL_COMPLETION ||
+		    kurb->status == USBD_SHORT_XFER) {
+			ur = &kurb->req;
+			len = UGETW(ur->ucr_request.wLength);
+			if (len > xfer->actlen)
+				len = xfer->actlen;
+			kurb->actlen = len;
+			if (len != 0) {
+				iov.iov_base = (caddr_t)ur->ucr_data;
+				iov.iov_len = len;
+				uio.uio_iov = &iov;
+				uio.uio_iovcnt = 1;
+				uio.uio_resid = len;
+				uio.uio_offset = 0;
+				uio.uio_segflg = UIO_USERSPACE;
+				uio.uio_rw =
+					ur->ucr_request.bmRequestType & UT_READ ?
+					UIO_READ : UIO_WRITE;
+				uio.uio_procp = p;
+				if (uio.uio_rw == UIO_READ) {
+					error = uiomove(KERNADDR(&xfer->dmabuf, 0), len, &uio);
+					if (error) {
+						usbd_free_xfer(xfer);
+						free(kurb, M_TEMP, sizeof(*kurb));
+						return (error);
+					}
+				}
 			}
 		}
-	ret:
-		if (ptr)
-			free(ptr, M_TEMP, 0);
-		return (error);
+		*urb = *kurb;
+		usbd_free_xfer(kurb->xfer);
+		free(kurb, M_TEMP, sizeof(*kurb));
+		break;
 	}
 	case USB_GET_DEVICEINFO:
 		usbd_fill_deviceinfo(sc->sc_udev,
@@ -1248,44 +1593,54 @@
 	if (sce == NULL)
 		return (POLLERR);
 #ifdef DIAGNOSTIC
-	if (!sce->edesc) {
-		printf("ugenpoll: no edesc\n");
-		return (POLLERR);
+	if (UGENENDPOINT(dev) != USB_CONTROL_ENDPOINT) {
+		if (!sce->edesc) {
+			printf("ugenpoll: no edesc\n");
+			return (POLLERR);
+		}
+		if (!sce->pipeh) {
+			printf("ugenpoll: no pipe\n");
+			return (POLLERR);
+		}
 	}
-	if (!sce->pipeh) {
-		printf("ugenpoll: no pipe\n");
-		return (POLLERR);
-	}
 #endif
 	s = splusb();
-	switch (sce->edesc->bmAttributes & UE_XFERTYPE) {
-	case UE_INTERRUPT:
+	if (UGENENDPOINT(dev) == USB_CONTROL_ENDPOINT) {
 		if (events & (POLLIN | POLLRDNORM)) {
-			if (sce->q.c_cc > 0)
+			if (!TAILQ_EMPTY(&sce->queue))
 				revents |= events & (POLLIN | POLLRDNORM);
 			else
 				selrecord(p, &sce->rsel);
 		}
-		break;
-	case UE_ISOCHRONOUS:
-		if (events & (POLLIN | POLLRDNORM)) {
-			if (sce->cur != sce->fill)
-				revents |= events & (POLLIN | POLLRDNORM);
-			else
-				selrecord(p, &sce->rsel);
+	} else {
+		switch (sce->edesc->bmAttributes & UE_XFERTYPE) {
+		case UE_INTERRUPT:
+			if (events & (POLLIN | POLLRDNORM)) {
+				if (sce->q.c_cc > 0)
+					revents |= events & (POLLIN | POLLRDNORM);
+				else
+					selrecord(p, &sce->rsel);
+			}
+			break;
+		case UE_ISOCHRONOUS:
+			if (events & (POLLIN | POLLRDNORM)) {
+				if (sce->cur != sce->fill)
+					revents |= events & (POLLIN | POLLRDNORM);
+				else
+					selrecord(p, &sce->rsel);
+			}
+			break;
+		case UE_BULK:
+			if (events & (POLLIN | POLLRDNORM)) {
+				if (!TAILQ_EMPTY(&sce->queue))
+					revents |= events & (POLLIN | POLLRDNORM);
+				else
+					selrecord(p, &sce->rsel);
+			}
+			break;
+		default:
+			break;
 		}
-		break;
-	case UE_BULK:
-		/*
-		 * We have no easy way of determining if a read will
-		 * yield any data or a write will happen.
-		 * Pretend they will.
-		 */
-		revents |= events &
-			   (POLLIN | POLLRDNORM | POLLOUT | POLLWRNORM);
-		break;
-	default:
-		break;
 	}
 	splx(s);
 	return (revents);
