--- usb.h.orig	Thu Jun  4 13:37:23 2015
+++ usb.h	Tue Jun 30 17:39:25 2015
@@ -41,6 +41,8 @@
 
 #include <sys/ioctl.h>
 
+#include <sys/queue.h>
+
 #define USB_STACK_VERSION 2
 
 #define USB_MAX_DEVICES 128
@@ -612,6 +614,17 @@
 	int	ucr_actlen;		/* actual length transferred */
 };
 
+struct ctl_urb {
+	struct usb_ctl_request req;
+	void *xfer;
+	int actlen;
+	void *sce;
+	void *dmabuf;
+	int status;
+	TAILQ_ENTRY(ctl_urb) entries;
+	void *user_context;
+};
+
 struct usb_alt_interface {
 	int	uai_config_index;
 	int	uai_interface_index;
@@ -748,10 +761,11 @@
 #define USB_GET_ENDPOINT_DESC	_IOWR('U', 108, struct usb_endpoint_desc)
 #define USB_GET_FULL_DESC	_IOWR('U', 109, struct usb_full_desc)
 #define USB_GET_STRING_DESC	_IOWR('U', 110, struct usb_string_desc)
-#define USB_DO_REQUEST		_IOWR('U', 111, struct usb_ctl_request)
+#define USB_DO_REQUEST		_IOWR('U', 111, struct ctl_urb)
 #define USB_GET_DEVICEINFO	_IOR ('U', 112, struct usb_device_info)
 #define USB_SET_SHORT_XFER	_IOW ('U', 113, int)
 #define USB_SET_TIMEOUT		_IOW ('U', 114, int)
+#define USB_GET_COMPLETED       _IOWR('U', 115, struct ctl_urb)
 
 /* Modem device */
 #define USB_GET_CM_OVER_DATA	_IOR ('U', 130, int)
