--- usb.h.orig	Thu Jun  4 13:37:23 2015
+++ usb.h	Wed Jun 10 14:00:19 2015
@@ -749,6 +749,7 @@
 #define USB_GET_FULL_DESC	_IOWR('U', 109, struct usb_full_desc)
 #define USB_GET_STRING_DESC	_IOWR('U', 110, struct usb_string_desc)
 #define USB_DO_REQUEST		_IOWR('U', 111, struct usb_ctl_request)
+#define USB_GET_COMPLETED	_IOWR('U', XXX, struct usb_ctl_request)
 #define USB_GET_DEVICEINFO	_IOR ('U', 112, struct usb_device_info)
 #define USB_SET_SHORT_XFER	_IOW ('U', 113, int)
 #define USB_SET_TIMEOUT		_IOW ('U', 114, int)
